keystore/src/connection/platform/wasm/migration.rs

```diff
@@ -40,23 +50,92 @@ pub(crate) async fn migrate(from: u32, final_target: u32, name: &str, key: &str)
+    let mut proteus_prekeys = ProteusPrekey::convert_from_v_1_2_0_or_earlier(&mut old_connection).await?;
+    let mut proteus_identities = ProteusIdentity::convert_from_v_1_2_0_or_earlier(&mut old_connection).await?;
+    let mut proteus_sessions = ProteusSession::convert_from_v_1_2_0_or_earlier(&mut old_connection).await?;
+    // Fetching old records finished.
+    drop(old_connection);
     old_storage.close().await?;
 
-    // Now store all converted records in the new storage.
-    // This will overwrite all previous entities in the DB.
-    // Cannot use public API here because we would end in a never-ending loop
+    // Create new storage. Cannot use public API here because we would end in a never-ending loop
     let new_idb = get_builder(name, version).build().await?;
     let new_wrapper = WasmStorageWrapper::Persistent(new_idb);
     let mut new_storage = WasmEncryptedStorage::new(key, new_wrapper);
-
+    // Now store all converted records in the new storage.
+    // This will overwrite all previous entities in the DB.
     new_storage
         .save(MlsCredential::COLLECTION_NAME, &mut credentials)
         .await?;
-
+    new_storage
+        .save(MlsSignatureKeyPair::COLLECTION_NAME, &mut signature_keys)
+        .await?;
+    new_storage
+        .save(MlsHpkePrivateKey::COLLECTION_NAME, &mut hpke_keys)
+        .await?;
+    new_storage
+        .save(MlsEncryptionKeyPair::COLLECTION_NAME, &mut encryption_keys)
+        .await?;
+    new_storage
+        .save(MlsEpochEncryptionKeyPair::COLLECTION_NAME, &mut epoch_encryption_keys)
+        .await?;
+    new_storage
+        .save(MlsPskBundle::COLLECTION_NAME, &mut psk_bundles)
+        .await?;
+    new_storage
+        .save(MlsKeyPackage::COLLECTION_NAME, &mut key_packages)
+        .await?;
+    new_storage
+        .save(PersistedMlsGroup::COLLECTION_NAME, &mut groups)
+        .await?;
+    new_storage
+        .save(PersistedMlsPendingGroup::COLLECTION_NAME, &mut pending_groups)
+        .await?;
+    new_storage
+        .save(MlsPendingMessage::COLLECTION_NAME, &mut pending_messages)
+        .await?;
+    new_storage
+        .save(E2eiEnrollment::COLLECTION_NAME, &mut e2ei_enrollments)
+        .await?;
+    new_storage
+        .save(E2eiRefreshToken::COLLECTION_NAME, &mut e2ei_tokens)
+        .await?;
+    new_storage
+        .save(E2eiAcmeCA::COLLECTION_NAME, &mut e2ei_acme_cas)
+        .await?;
+    new_storage
+        .save(E2eiIntermediateCert::COLLECTION_NAME, &mut e2ei_intermediates)
+        .await?;
+    new_storage.save(E2eiCrl::COLLECTION_NAME, &mut e2ei_crls).await?;
+    new_storage
+        .save(ProteusPrekey::COLLECTION_NAME, &mut proteus_prekeys)
+        .await?;
+    new_storage
+        .save(ProteusIdentity::COLLECTION_NAME, &mut proteus_identities)
+        .await?;
+    new_storage
+        .save(ProteusSession::COLLECTION_NAME, &mut proteus_sessions)
+        .await?;
+    // Migration finished
     new_storage.close()?;
```

Comment:
Is this transactional? What guarantees do we have that if there is an error in the middle of the above block of code, that database content is still valid, i.e. not partially migrated?

Reply:
We can now recover from partial migration caused by interruptions, everything else will be handled in WPB-10934